### ISSUE TITLE
[Fuzzer] Testcase "fuzzilli('FUZZILLI_CRASH', {0 | 1 | 2})" did not crash

### ISSUE THREAD

**Issue by extf33** [googleprojectzero/fuzzilli#391]
Fuzzilli: [`8eec7a7`](https://github.com/googleprojectzero/fuzzilli/commit/8eec7a70fc49ed49f92cec9a29a42cbf17764638)
V8: `92a918e10bd36c1045b2f750b56fdab4b4148ae4`

I've got
```
[Fuzzer] Testcase "fuzzilli('FUZZILLI_CRASH', 0)" did not crash
```
```
[Fuzzer] Testcase "fuzzilli('FUZZILLI_CRASH', 1)" did not crash
```
```
[Fuzzer] Testcase "fuzzilli('FUZZILLI_CRASH', 2)" did not crash
```
in release, debug build.

**Comment by saelo** [comment]
Hi! Did you compile V8 [like this](https://github.com/googleprojectzero/fuzzilli/blob/main/Targets/V8/fuzzbuild.sh)?

**Comment by saelo** [comment]
Have you verified that e.g. `fuzzilli('FUZZILLI_CRASH', 1)` actually crashes in that d8 binary? You could also try running that d8 binary with the REPRLRun utility: `swift run REPRLRun path/to/d8` and check if the crashes are detected there.

**Comment by xie-22** [comment]
@extf33 @saelo  I encountered the same problem today. I added `is_asan=true` to build.sh, but it still wouldn't run. I discovered that it was because asan wasn't given enough time to initialize. So, Fuzzilli setting `--timeout = 1500` or a longer value solved the problem.

**Comment by xie-22** [comment]
## log
### Fail
```bash
root@xie22:/v8/fuzzilli# swift run -c release FuzzilliCli   --profile=v8   --jobs=8   --timeout=12000   --stor
agePath=/v8/fuzz_results   --overwrite   /v8/out/fuzzbuild/d8
Building for production...
[1/1] Write swift-version--42C19770CB19FCAE.txt
Build of product 'FuzzilliCli' complete! (0.23s)

// other code 
lainTime", "withResolvers", "withTimeZone", "xor", "zip", "zipKeyed", "zonedDateTimeISO"]
[JavaScriptEnvironment] Have 8 custom property names: ["!", "42", "a", "b", "c", "d", "e", "f"]
[JavaScriptEnvironment] Have 7 custom method names: ["67", "?", "m", "n", "o", "toString", "valueOf"]
[JavaScriptEnvironment] Have 4 custom private method names: ["m", "n", "o", "p"]
[JavaScriptEnvironment] Have 6 custom identifier names: ["a", "b", "c", "d", "e", "f"]
[Fuzzer] Changing state from uninitialized to corpusGeneration
[Fuzzer] Initialized
[Fuzzer] Testcase "fuzzilli('FUZZILLI_CRASH', 0)" timed out, the configured timeout threshold (12000ms) might be too low
[Fuzzer] Shutting down due to fatal error

++++++++++ Fuzzer Finished ++++++++++

Fuzzer Statistics
-----------------
Fuzzer state:                 Initial corpus generation (with GenerativeEngine)
Uptime:                       0d 0h 0m 13s
Total Samples:                0
Interesting Samples Found:    0
Last Interesting Sample:      0d 0h 0m 13s
Valid Samples Found:          0
Corpus Size:                  0
Correctness Rate:             0.00% (overall: 0.00%)
Timeout Rate:                 0.00% (overall: 0.00%)
Crashes Found:                0
Timeouts Hit:                 0
Coverage:                     0.00%
Avg. program size:            0.00
Avg. corpus program size:     0.00
Avg. program execution time:  13ms
Connected nodes:              0
Execs / Second:               0.00
Fuzzer Overhead:              14.59%
Minimization Overhead:        0.00%
Total Execs:                  9

Aborting execution after a fatal error.
```
### Success
```bash
root@xie22:/v8/fuzzilli# swift run -c release FuzzilliCli   --profile=v8   --jobs=8   --timeout=15000   --stor
agePath=/v8/fuzz_results   --overwrite   /v8/out/fuzzbuild/d8
Building for production...
[1/1] Write swift-version--42C19770CB19FCAE.txt
Build of product 'FuzzilliCli' complete! (0.22s)
[Cli] Using the following arguments for the target engine: ["--expose-gc", "--expose-externalize-string", "--omit-quit", "--allow-natives-syntax", "--fuzzing", "--harmony", "--experimental-fuzzing", "--js-staging", "--wasm-staging", "--wasm-fast-api", "--expose-fast-api", "--wasm-test-streaming"]
[Cli] Enabled mutators: ["ExplorationMutator", "CodeGenMutator", "SpliceMutator", "ProbingMutator", "InputMutator (loose)", "InputMutator (aware)", "OperationMutator", "CombineMutator"]
[Cli] Loaded additional builtins from profile: ["createExternalizableTwoByteString", "gc", "Worker", "d8", "createExternalizableString", "externalizeString", "isOneByteString"]
[Cli] Loaded additional ObjectGroups from profile: ["D8", "D8Test", "D8FastCAPI", "GCOptions", "Worker", "Worker.prototype", "WorkerConstructor"]
[Cli] Loaded additional Enumerations from profile: ["gcType", "gcExecution"]
[Cli] Deleting all files in /v8/fuzz_results due to --overwrite
[Coverage] Initialized, 1877242 edges
[JavaScriptEnvironment] Checking constructor availability...

// othe code

lainTime", "withResolvers", "withTimeZone", "xor", "zip", "zipKeyed", "zonedDateTimeISO"]
[JavaScriptEnvironment] Have 8 custom property names: ["!", "42", "a", "b", "c", "d", "e", "f"]
[JavaScriptEnvironment] Have 7 custom method names: ["67", "?", "m", "n", "o", "toString", "valueOf"]
[JavaScriptEnvironment] Have 4 custom private method names: ["m", "n", "o", "p"]
[JavaScriptEnvironment] Have 6 custom identifier names: ["a", "b", "c", "d", "e", "f"]
[Fuzzer] Changing state from uninitialized to corpusGeneration
[Fuzzer] Initialized
[Fuzzer] Recommended timeout: at least 24840ms. Current timeout: 15000ms
[Fuzzer] Startup tests finished successfully
[Fuzzer] Let's go!
[Cli] Enabled mutators: ["ExplorationMutator", "CodeGenMutator", "SpliceMutator", "ProbingMutator", "InputMutator (loose)", "InputMutator (aware)", "OperationMutator", "CombineMutator"]
[Cli] Loaded additional builtins from profile: ["createExternalizableTwoByteString", "gc", "Worker", "d8", "createExternalizableString", "externalizeString", "isOneByteString"]
[Cli] Loaded additional ObjectGroups from profile: ["D8", "D8Test", "D8FastCAPI", "GCOptions", "Worker", "Worker.prototype", "WorkerConstructor"]
[Cli] Loaded additional Enumerations from profile: ["gcType", "gcExecution"]
[Cli] Enabled mutators: ["ExplorationMutator", "CodeGenMutator", "SpliceMutator", "ProbingMutator", "InputMutator (loose)", "InputMutator (aware)", "OperationMutator", "CombineMutator"]
[Cli] Loaded additional builtins from profile: ["createExternalizableTwoByteString", "gc", "Worker", "d8", "createExternalizableString", "externalizeString", "isOneByteString"]
[Cli] Loaded additional ObjectGroups from profile: ["D8", "D8Test", "D8FastCAPI", "GCOptions", "Worker", "Worker.prototype", "WorkerConstructor"]
[Cli] Loaded additional Enumerations from profile: ["gcType", "gcExecution"]
[Cli] Enabled mutators: ["ExplorationMutator", "CodeGenMutator", "SpliceMutator", "ProbingMutator", "InputMutator (loose)", "InputMutator (aware)", "OperationMutator", "CombineMutator"]
[Cli] Loaded additional builtins from profile: ["createExternalizableTwoByteString", "gc", "Worker", "d8", "createExternalizableString", "externalizeString", "isOneByteString"]
[Cli] Loaded additional ObjectGroups from profile: ["D8", "D8Test", "D8FastCAPI", "GCOptions", "Worker", "Worker.prototype", "WorkerConstructor"]
[Cli] Loaded additional Enumerations from profile: ["gcType", "gcExecution"]
[Cli] Enabled mutators: ["ExplorationMutator", "CodeGenMutator", "SpliceMutator", "ProbingMutator", "InputMutator (loose)", "InputMutator (aware)", "OperationMutator", "CombineMutator"]
[Cli] Loaded additional builtins from profile: ["createExternalizableTwoByteString", "gc", "Worker", "d8", "createExternalizableString", "externalizeString", "isOneByteString"]
[Cli] Loaded additional ObjectGroups from profile: ["D8", "D8Test", "D8FastCAPI", "GCOptions", "Worker", "Worker.prototype", "WorkerConstructor"]
[Cli] Loaded additional Enumerations from profile: ["gcType", "gcExecution"]
[Cli] Enabled mutators: ["ExplorationMutator", "CodeGenMutator", "SpliceMutator", "ProbingMutator", "InputMutator (loose)", "InputMutator (aware)", "OperationMutator", "CombineMutator"]
[Cli] Loaded additional builtins from profile: ["createExternalizableTwoByteString", "gc", "Worker", "d8", "createExternalizableString", "externalizeString", "isOneByteString"]
[Cli] Loaded additional ObjectGroups from profile: ["D8", "D8Test", "D8FastCAPI", "GCOptions", "Worker", "Worker.prototype", "WorkerConstructor"]
[Cli] Loaded additional Enumerations from profile: ["gcType", "gcExecution"]
[Cli] Enabled mutators: ["ExplorationMutator", "CodeGenMutator", "SpliceMutator", "ProbingMutator", "InputMutator (loose)", "InputMutator (aware)", "OperationMutator", "CombineMutator"]
[Cli] Loaded additional builtins from profile: ["createExternalizableTwoByteString", "gc", "Worker", "d8", "createExternalizableString", "externalizeString", "isOneByteString"]
[Cli] Loaded additional ObjectGroups from profile: ["D8", "D8Test", "D8FastCAPI", "GCOptions", "Worker", "Worker.prototype", "WorkerConstructor"]
[Cli] Loaded additional Enumerations from profile: ["gcType", "gcExecution"]
[Cli] Enabled mutators: ["ExplorationMutator", "CodeGenMutator", "SpliceMutator", "ProbingMutator", "InputMutator (loose)", "InputMutator (aware)", "OperationMutator", "CombineMutator"]
[Cli] Loaded additional builtins from profile: ["createExternalizableTwoByteString", "gc", "Worker", "d8", "createExternalizableString", "externalizeString", "isOneByteString"]
[Cli] Loaded additional ObjectGroups from profile: ["D8", "D8Test", "D8FastCAPI", "GCOptions", "Worker", "Worker.prototype", "WorkerConstructor"]
[Cli] Loaded additional Enumerations from profile: ["gcType", "gcExecution"]
Fuzzer Statistics
-----------------
Fuzzer state:                 Initial corpus generation (with GenerativeEngine)
Uptime:                       0d 0h 1m 3s
Total Samples:                17
Interesting Samples Found:    13
Last Interesting Sample:      0d 0h 0m 0s
Valid Samples Found:          14
Corpus Size:                  13
Correctness Rate:             82.35% (overall: 82.35%)
Timeout Rate:                 0.00% (overall: 0.00%)
Crashes Found:                0
Timeouts Hit:                 0
Coverage:                     1.29%
Avg. program size:            44.47
Avg. corpus program size:     24.46
Avg. program execution time:  15ms
Connected nodes:              0
Execs / Second:               56.06
Fuzzer Overhead:              4.32%
Minimization Overhead:        90.05%
Total Execs:                  1176

Fuzzer Statistics
-----------------
Fuzzer state:                 Initial corpus generation (with GenerativeEngine)
Uptime:                       0d 0h 2m 4s
Total Samples:                40
Interesting Samples Found:    28
Last Interesting Sample:      0d 0h 0m 0s
Valid Samples Found:          31
Corpus Size:                  27
Correctness Rate:             77.50% (overall: 77.50%)
Timeout Rate:                 2.50% (overall: 2.50%)
Crashes Found:                0
Timeouts Hit:                 1
Coverage:                     2.15%
Avg. program size:            42.20
Avg. corpus program size:     21.36
Avg. program execution time:  14ms
Connected nodes:              0
Execs / Second:               23.22
Fuzzer Overhead:              4.81%
Minimization Overhead:        89.38%
Total Execs:                  2091

[Corpus] Successfully serialized 33 programs
```

**Comment by Liedtke** [comment]
So from a Fuzzilli perspective this is working as intended, the startup test cases try to flush out any issues and if they do, they abort the run.

> [Fuzzer] Testcase "fuzzilli('FUZZILLI_CRASH', 0)" timed out, the configured timeout threshold (12000ms) might be too low

We changed the reporting here some months ago and it now explains what's going wrong, I don't think there is anything else actionable?
Though it sounds surprising that it needs a > 12 seconds timeout?
When I run it locally on a Linux workstation with an ASAN V8 build, it uses a default timeout of 500-600 ms and the startup tests pass without issues. (This is with empty `ASAN_OPTIONS`.)

**Comment by xie-22** [comment]
@Liedtke hi,Thank you for your reply.

> Testing revealed that running `--resume` based on a previous corpus can cause timeout crashes when importing large corpora. Adding `--timeout=60000` will alleviate this issue. 

After thoroughly researching and understanding the v8 vulnerability, I completely rewrote the solution and added the following(**This may only represent my personal experience with issues encountered during rewriting and crashes encountered without rewriting, impacting stability, coverage, fuzzing, inability to effectively record the total number of crashes, inability to effectively discover vulnerabilities, and negative results, etc.)**:


----

### Summary of Changes

This PR resolves several critical stability, coverage, and performance bottlenecks identified during long-running fuzzing campaigns on V8 (Turboshaft/Maglev):

#### 1. Instant Crash Detection & WSL2 `core_pattern` Fix
* **Root Cause**: On WSL2 environments, `core_pattern` defaults to piping core dumps (`|/wsl-capture-crash`), causing ASAN-instrumented processes to hang for seconds upon crashing. REPRL misclassified real crashes as timeouts.
* **Fix**: Added `prctl(PR_SET_DUMPABLE, 0)` in `libreprl-posix.c` and updated campaign runners to enforce `kernel.core_pattern=/dev/null`. Startup crash test detection improved from ~1000ms (timeout) to 36-39ms, successfully re-enabling all startup assertions.

#### 2. Robust Scope & FuzzIL Static Verification
* **Fix**: Resolved `codeVerificationError` ("variable not visible / invalid context") across `CodeGenMutator`, `PropertyAccessorMutator`, and `ProbingMutator` by enforcing strict active block-scope lifetime tracking and context requirement validation. Invalid AST mutations are now rejected defensively without leaking out-of-scope variables.

#### 3. Corpus Resilience & Deserialization Safety
* **Fix**: Implemented a "Discard and Continue" sandbox pattern in Protobuf/Corpus deserialization. Corrupted, legacy, or schema-mismatched `.fzil` samples are safely logged and discarded rather than triggering fatal unwrapping assertions.

#### 4. Malformed Unicode & String Subsystem Engine
* **Feature**: Introduced `MalformedStringGenerator` with a curated pool of 44 high-risk string patterns (lone surrogates encoded safely via JS escape sequences, zero-width/Bidi controls, case-folding expansions, ZWJ clusters). Targeted at V8 string flattening, `ConsString` transcoding, and RegExp JIT boundaries.

#### 5. Private Property Generator Fallbacks
* **Fix**: Fixed 0.00% invocation success rates for `PrivateProperty*` generators by adding self-healing inline class expression fallbacks when no matching instance exists in the current scope.

#### 6. Anti-hang Complexity Hard Limits & Energy Scheduling
* **Optimization**: Enforced `maxProgramSize=2000` and `maxBlockNestingDepth=64` to prune excessive loop nesting during splicing. Switched to `EnergyCorpus` scheduling, reducing average execution time from ~170ms to ~35-42ms and boosting throughput by over 3x (to 60-70+ exec/s).

---

### Test & Verification
* `swift build -c release`: **0 errors, 0 warnings**
* `swift test`: **739/739 tests passed (36 suites)** including `MalformedStringTests`
* Verified on V8 15.3 (Turboshaft/Maglev assert builds).

----

**Comment by xie-22** [comment]
This is my updated FuzzilliCli. :)

----

```bash
root@xie22:~# FuzzilliCli --help
Usage:
FuzzilliCli [options] --profile=<profile> /path/to/jsshell

Options:
    --profile=name               : Select one of several preconfigured profiles.
                                   Available profiles: ["v8holefuzzing", "njs", "duktape", "qtjs", "v8Dumpling", "xs", "v8", "v8Sandbox", "qjs", "jsc", "spidermonkey", "v8Differential", "jerryscript", "serenity"].
    --jobs=n                     : Total number of fuzzing jobs. This will start a main instance and n-1 worker instances.
    --engine=name                : The fuzzing engine to use. Available engines: "mutation" (default), "hybrid", "multi".
                                   Only the mutation engine should be regarded stable at this point.
    --corpus=name                : The corpus scheduler to use. Available schedulers: "basic" (default), "markov",
                                   "energy" (AFL++-style energy allocation)
    --logLevel=level             : The log level to use. Valid values: "verbose", "info", "warning", "error", "fatal" (default: "info").
    --maxIterations=n            : Run for the specified number of iterations (default: unlimited).
    --maxRuntimeInHours=n        : Run for the specified number of hours (default: unlimited).
    --timeout=n                  : Timeout in ms after which to interrupt execution of programs (default depends
                                   on the profile). Or provide an interval like --timeout=200,400. The actual
                                   timeout in this interval will be determined by the start-up tests.
    --minMutationsPerSample=n    : Discard samples from the corpus only after they have been mutated at least this many times (default: 25).
    --minCorpusSize=n            : Keep at least this many samples in the corpus regardless of the number of times
                                   they have been mutated (default: 1000).
    --maxCorpusSize=n            : Only allow the corpus to grow to this many samples. Otherwise the oldest samples
                                   will be discarded (default: unlimited).
    --markovDropoutRate=p        : Rate at which low edge samples are not selected, in the Markov Corpus Scheduler,
                                   per round of sample selection. Used to ensure diversity between fuzzer instances
                                   (default: 0.10)
    --consecutiveMutations=n     : Perform this many consecutive mutations on each sample (default: 5).
    --corpusGenerationIterations=n : Switch from corpus generation to the main fuzzing phase after this many
                                   iterations without finding a new interesting sample (default: 100).
    --minimizationLimit=p        : When minimizing interesting programs, keep at least this percentage of the original instructions
                                   regardless of whether they are needed to trigger the interesting behaviour or not.
                                   See Minimizer.swift for an overview of this feature (default: 0.0).
    --storagePath=path           : Path at which to store output files (crashes, corpus, etc.) to.
    --resume                     : If storage path exists, import the programs from the corpus/ subdirectory
    --overwrite                  : If storage path exists, delete all data in it and start a fresh fuzzing session
    --staticCorpus               : In this mode, we will just mutate the existing corpus and look for crashes.
                                   No new samples are added to the corpus, regardless of their coverage.
                                   This can be used to find different manifestations of bugs and
                                   also to try and reproduce a flaky crash or turn it into a deterministic one.
    --exportStatistics           : If enabled, fuzzing statistics will be collected and saved to disk in regular intervals.
                                   Requires --storagePath.
    --statisticsExportInterval=n : Interval in minutes for saving fuzzing statistics to disk (default: 10).
                                   Requires --exportStatistics.
    --importCorpus=path          : Imports an existing corpus of FuzzIL programs to build the initial corpus for fuzzing.
                                   The provided path must point to a directory, and all .fzil files in that directory will be imported.
    --corpusImportMode=mode      : The corpus import mode. Possible values:
                                             default : Keep samples that are interesting (e.g. those that increase code coverage) and minimize them (default).
                                                full : Keep all samples that execute successfully without minimization.
                                         unminimized : Keep samples that are interesting but do not minimize them.

    --instanceType=type          : Specifies the instance type for distributed fuzzing over a network.
                                   In distributed fuzzing, instances form a tree hierarchy, so the possible values are:
                                               root: Accept connections from other instances.
                                               leaf: Connect to a parent instance and synchronize with it.
                                       intermediate: Connect to a parent instance and synchronize with it but also accept incoming connections.
                                         standalone: Don't participate in distributed fuzzing (default).
                                   Note: it is *highly* recommended to run distributed fuzzing in an isolated network!
    --bindTo=host:port           : When running as a root or intermediate node, bind to this address (default: 127.0.0.1:1337).
    --connectTo=host:port        : When running as a leaf or intermediate node, connect to the parent instance at this address (default: 127.0.0.1:1337).
    --corpusSyncMode=mode        : How the corpus is synchronized during distributed fuzzing. Possible values:
                                                  up: newly discovered corpus samples are only sent to parent nodes but
                                                      not to chjild nodes. This way, the child nodes are forced to generate their
                                                      own corpus, which may lead to more diverse samples overall. However, parent
                                                      instances will still have the full corpus.
                                                down: newly discovered corpus samples are only sent to child nodes but not to
                                                      parent nodes. This may make sense when importing a corpus in the parent.
                                      full (default): newly discovered corpus samples are sent in both direction. This is the
                                                      default behaviour and will generally cause all instances in the network
                                                      to have very roughly the same corpus.
                                               none : corpus samples are not shared with any other instances in the network.
                                   Note: thread workers (--jobs=X) always fully synchronize their corpus.
    --diagnostics                : Enable saving of programs that failed or timed-out during execution. Also tracks
                                   executions on the current REPRL instance.
    --swarmTesting               : Enable Swarm Testing mode. The fuzzer will choose random weights for the code generators per process.
    --inspect                    : Enable inspection for generated programs. When enabled, additional .fuzzil.history files are written
                                   to disk for every interesting or crashing program. These describe in detail how the program was generated
                                   through mutations, code generation, and minimization.
    --argumentRandomization      : Enable JS engine argument randomization
    --additionalArguments=args   : Pass additional arguments to the JS engine. If multiple arguments are passed, they should be separated by a comma.
    --tag=tag                    : Optional string tag associated with this instance which will be stored in the settings.json file as well as in crashing samples.
                                   This can for example be used to remember the target revision that is being fuzzed.
    --wasm                       : Enable Wasm CodeGenerators (see WasmCodeGenerators.swift).
    --enable-wasm                : Alias for --wasm.
    --wasm-opt-path=path         : Path to the wasm-opt binary to enable Binaryen Wasm generation.
    --forDifferentialFuzzing     : Enable additional features for better support of external differential fuzzing.
    --bundle                     : Generate bundles containing multiple JS scripts and modules
    --skip-startup-tests         : Skip the startup crash/timeout tests. Useful when running under
                                   sanitizers whose signal handling interferes with the tests.
    --fast-start                 : Alias for --skip-startup-tests.
    --reproducibilityRuns=n      : Consecutive reproducible executions required to register a
                                   crash or differential finding as deterministic (default: 3).

Conflicting or incompatible options do not abort the process: a warning is
printed and a safe default is used instead (e.g. --resume wins over --overwrite,
--maxRuntimeInHours wins over --maxIterations, and corpus-incompatible size
options fall back to the basic corpus).
```

----
